libsql-sqlite3/src/vectordiskann.c

```diff
@@ -0,0 +1,345 @@
+  rc = sqlite3_blob_read(pBlobSpot->pBlob, pBlobSpot->pBuffer, nBufferSize, 0);
+  if( rc != SQLITE_OK ){
+    return rc;
+  }
+  pBlobSpot->isInitialized = 1;
+  return SQLITE_OK;
+}
+
+int blobSpotFlush(BlobSpot *pBlobSpot) {
+  return sqlite3_blob_write(pBlobSpot->pBlob, pBlobSpot->pBuffer, pBlobSpot->nBufferSize, 0);
+}
+
+void blobSpotFree(BlobSpot *pBlobSpot) {
+  if( pBlobSpot->pBlob != NULL ){
+    sqlite3_blob_close(pBlobSpot->pBlob);
+  }
+  if( pBlobSpot->pBuffer != NULL ){
+    sqlite3_free(pBlobSpot->pBuffer);
+  }
+  sqlite3_free(pBlobSpot);
+}
+
+/**************************************************************************
+** Layout specific utilities
+**************************************************************************/
+
+int nodeEdgesMaxCount(const DiskAnnIndex *pIndex){
+  unsigned int nMaxEdges = (pIndex->nBlockSize - pIndex->nNodeVectorSize - VECTOR_NODE_METADATA_SIZE) / (pIndex->nEdgeVectorSize + VECTOR_EDGE_METADATA_SIZE);
+  assert( nMaxEdges > 0);
+  return nMaxEdges;
+}
+
+int nodeEdgesMetadataOffset(const DiskAnnIndex *pIndex){
+  unsigned int offset;
+  unsigned int nMaxEdges = nodeEdgesMaxCount(pIndex);
+  offset = VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + nMaxEdges * pIndex->nEdgeVectorSize;
+  assert( offset <= pIndex->nBlockSize );
+  return offset;
+}
+
+void nodeBinInit(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, u64 nRowid, Vector *pVector){
+  assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize <= pBlobSpot->nBufferSize );
+
+  memset(pBlobSpot->pBuffer, 0, pBlobSpot->nBufferSize);
+  writeLE64(pBlobSpot->pBuffer, nRowid);
+  writeLE16(pBlobSpot->pBuffer + sizeof(u64), 0);
+  vectorSerializeToBlob(pVector, pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE, pIndex->nNodeVectorSize);
+}
+
+void nodeBinVector(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot, Vector *pVector) {
+  assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize <= pBlobSpot->nBufferSize );
+
+  vectorInitStatic(pVector, pIndex->nNodeVectorType, pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE, pIndex->nNodeVectorSize);
+}
+
+u16 nodeBinEdges(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot) {
+  assert( VECTOR_NODE_METADATA_SIZE <= pBlobSpot->nBufferSize );
+
+  return readLE16(pBlobSpot->pBuffer + sizeof(u64));
+}
+
+void nodeBinEdge(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot, int iEdge, u64 *pRowid, Vector *pVector) {
+  int offset = nodeEdgesMetadataOffset(pIndex);
+
+  if( pRowid != NULL ){
+    assert( offset + (iEdge + 1) * VECTOR_EDGE_METADATA_SIZE <= pBlobSpot->nBufferSize );
+    *pRowid = readLE64(pBlobSpot->pBuffer + offset + iEdge * VECTOR_EDGE_METADATA_SIZE + sizeof(u64));
+  }
+  if( pVector != NULL ){
+    assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + iEdge * pIndex->nEdgeVectorSize < offset );
+    vectorInitStatic(
+      pVector,
+      pIndex->nEdgeVectorType,
+      pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE + (iEdge + 1) * pIndex->nNodeVectorSize,
```

Comment:
Shouldn't this be `pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + iEdge * pIndex->nEdgeVectorSize`? It seems that the assertion in line 249 is correct.

Reply:
Yes, another bug (missed that after introduction of separate edge and vector sizes)